ArduPlane/mode_takeoff.cpp

```diff
@@ -116,11 +115,28 @@ void ModeTakeoff::update()
             if (!plane.throttle_suppressed) {
                 gcs().send_text(MAV_SEVERITY_INFO, "Takeoff to %.0fm for %.1fm heading %.1f deg",
                                 alt, dist, direction);
+                plane.takeoff_state.start_time_ms = AP_HAL::millis();
                 takeoff_started = true;
             }
         }
     }
-
+    // check for optional takeoff timeout
+    if (plane.takeoff_state.start_time_ms != 0 && takeoff_started && plane.g2.takeoff_timeout > 0) {
+        const float ground_speed = AP::gps().ground_speed();
+        const float takeoff_min_ground_speed = 4;
+        if (ground_speed >= takeoff_min_ground_speed) {
+            plane.takeoff_state.start_time_ms = 0;
+        } else {
+            uint32_t now = AP_HAL::millis();
+            if (now - plane.takeoff_state.start_time_ms > (uint32_t)(1000U * plane.g2.takeoff_timeout)) {
+                gcs().send_text(MAV_SEVERITY_INFO, "Takeoff timeout: %.1f m/s speed < 4m/s", ground_speed);
+                plane.set_flight_stage(AP_FixedWing::FlightStage::NORMAL);
+                takeoff_started = false;
+                plane.arming.disarm(AP_Arming::Method::TAKEOFFTIMEOUT);
```

Comment:
So if the takeoff times out the AP will disarm? In the air?

Reply:
if you have not achieved 4m/s in the time, it disarms....this is currently used only for NAV_TAKEOFF and should also to MODE TAKEOFF...was just not included when the mode was constructed...